### ISSUE TITLE
Strange autocomplete bugs on Ubuntu, depending on space before variable

### ISSUE THREAD

**Issue by hofingermarkus** [spyder-ide/spyder#6700]
## Problem Description
This is a strange behavior that I only experience under Ubuntu:
Autocompletion very often doesn't work if I do not have a space between the function and the next variable i put into it  "function(var1.<NoTabCompletionHere>" . Sometimes it seems as this is also the case for  variables that follow a comma e.g. "var1,var2.<NoTabCompletionHere"
Also from time to time it seems as if the home key is broken in a sense that it doesn't jump before the 1st but before the 2nd character. This could be something else or these two bugs could be related...
I've encountered this behavior now on different machines, and different anaconda setups independently. 

### What steps will reproduce the problem?
The following doesn't bring up the tab completion on Ubuntu, but it works without any problem on Windows. 
``` python
import numpy as np
a = np.array([1,2,3])
np.exp(a.<NoTabCompletionHere>
```

Btw. adding an extra space before the variable name will show up the tab completion correctly:
``` python
import numpy as np
a = np.array([1,2,3])
np.exp(<Space>a.<TabCompletionWorksAgain>
```

## Version and main components

* Spyder Version: 3.2.6  (it seems to be a bug that exists for many months & versions by now.)
* Python Version: 3.5.4
* Qt Versions: 5.6.2, PyQt5 5.6 on Linux

## Dependencies
```
pyflakes >=0.6.0 :  1.6.0 (OK)
pycodestyle >=2.3:  2.3.1 (OK)
pygments >=2.0   :  2.2.0 (OK)
pandas >=0.13.1  :  0.21.1 (OK)
numpy >=1.7      :  1.14.1 (OK)
sphinx >=0.6.6   :  1.7.1 (OK)
rope >=0.9.4     :  0.10.7 (OK)
jedi >=0.9.0     :  0.11.1 (OK)
psutil >=0.3     :  5.4.3 (OK)
nbconvert >=4.0  :  5.3.1 (OK)
sympy >=0.7.3    :  1.1.1 (OK)
cython >=0.21    :  None (NOK)
qtconsole >=4.2.0:  4.3.1 (OK)
IPython >=4.0    :  6.2.1 (OK)
pylint >=0.25    :  1.8.2 (OK)

```

**Comment by bcolsen** [comment]
Unfortunately, I can't reproduce this in the editor on the latest anaconda on Ubuntu 16.04.

![peek 2018-03-09 15-17](https://user-images.githubusercontent.com/10513354/37232613-0baa6b88-23ad-11e8-932b-9dee830f91d3.gif)


Spyder version: 3.2.7 (works on master as well)

Python Version: 3.6.4

IPython >=4.0    :  6.2.1 (OK)
cython >=0.21    :  0.27.3 (OK)
jedi >=0.9.0     :  0.11.1 (OK)
nbconvert >=4.0  :  5.3.1 (OK)
numpy >=1.7      :  1.14.1 (OK)
pandas >=0.13.1  :  0.22.0 (OK)
psutil >=0.3     :  5.4.3 (OK)
pycodestyle >=2.3:  2.3.1 (OK)
pyflakes >=0.6.0 :  1.6.0 (OK)
pygments >=2.0   :  2.2.0 (OK)
pylint >=0.25    :  1.8.2 (OK)
qtconsole >=4.2.0:  4.3.1 (OK)
rope >=0.9.4     :  0.10.7 (OK)
sphinx >=0.6.6   :  1.7.1 (OK)
sympy >=0.7.3    :  1.1.1 (OK)

**Comment by CAM-Gerlach** [comment]
Thanks for reporting @hofingermarkus . Can you try with Spyder 3.2.7 and a more modern Python version and see if the bug still occurs?

**Comment by ccordoba12** [comment]
@csabella, what about you?

**Comment by csabella** [comment]
Inconclusive.  When I first tried it, autocompletion did not work, just as @hofingermarkus said.  So, I closed Spyder, reopened it in debug mode, and then autocompletion worked fine!  I have not had it 'not work' since, but I definitely saw the issue where `np.` was returning completions and `a.` was not.  One caveat though - when it wasn't working, even the space before the `a` didn't help me.

I plan on trying this every so often to see if I can detect a pattern.

**Comment by ccordoba12** [comment]
Thanks Cheryl! Let us know if/when you find a reproducible use case.

**Comment by csabella** [comment]
> I did my tests in spyders IPython Console (not the editor) and always get the same error

When I was able to recreate the issue, the IPython console within Spyder worked fine.  I had tested both just to make sure.  Many reports say that the Spyder editor isn't consistent with the Spyder IPython console, so I sort of expected it to work in the console and not the editor and didn't really pursue that further.

**Comment by bcolsen** [comment]
I can reproduce this in the spyder ipthon console, but not in the jupyter-qtconsole

spyder qtconsole

![spyder qtconsole](https://user-images.githubusercontent.com/10513354/37617895-03ddc452-2b7a-11e8-9efb-479c0791674b.gif)

jupyter qtconsole

![qtconsole](https://user-images.githubusercontent.com/10513354/37618008-5a85f202-2b7a-11e8-97f2-bbe37336ed73.gif)


The fuction tip is also a lot more usefull

**Comment by hofingermarkus** [comment]
**I think I found it!** It's a bug in the **Greedy Completer** - once it is turned on, the problems start!

How to reproduce:
- Reset Spyder Settings by removing the folder containing the settings files
- Start Spyder 
         => Test filename completion (works)
- Activate Greedy Compiler in the settings: (preferences->Ipyhton console -> Advanced Settings -> greedy completer
- Restart Ipython Console
-  Check again 
       => Filename completion is now broken (needs extra space), also triggers bug #6700 


![bugingreedycompleter](https://user-images.githubusercontent.com/12088794/37791334-407863fe-2e09-11e8-99f9-5ced954bbe51.gif)

**Comment by ccordoba12** [comment]
Thanks for finding the cause of this problem! Then this is a bug in IPython and not in Spyder. Unfortunately, the greedy completer is slated to be removed in a future IPython version, so they probably won't want to fix it.

**Comment by ccordoba12** [comment]
Latest IPython versions use Jedi to get completions like the one you posted above, but we had to deactivate that feature because it was making our console too slow when trying to get completions for large dataframes.

**Comment by hofingermarkus** [comment]
OK, 
Maybe a "quickfix" could be to inform the users in the settings dialog.
 1.) Update the Textmessage from the Greedy Completer with a passage to warn about the Ipython Bug.
 2.) Provide an option to activate Jedi in the Console, but also with a warning that this can slow down code completion for large dataframes.

**Comment by ccordoba12** [comment]
Please send us a PR against the 3.x branch for this. I don't think we'll have time to work on this right now.

**Comment by hofingermarkus** [comment]
Ok, I'll see If I find time this weekend...

**Comment by CAM-Gerlach** [comment]
Thanks @hofingermarkus !

**Comment by ccordoba12** [comment]
This was solved by PR #6832, but never referenced in a milestone.
